java/ql/src/Security/CWE/CWE-094/TemplateInjection.qll

```diff
@@ -0,0 +1,210 @@
+/** Definitions related to the Server Side Template Injection (SSTI) query. */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.FreeMarker
+import semmle.code.java.frameworks.Velocity
+import semmle.code.java.frameworks.JinJava
+import semmle.code.java.frameworks.Pebble
+import semmle.code.java.frameworks.Thymeleaf
+
+/** A taint tracking configuration to reason about Server Side Template Injection (SSTI) vulnerabilities */
+class TemplateInjectionFlowConfig extends TaintTracking::Configuration {
+  TemplateInjectionFlowConfig() { this = "TemplateInjectionFlowConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.getType() instanceof PrimitiveType or node.getType() instanceof BoxedType
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ) {
+    exists(AdditionalFlowStep a | a.isAdditionalTaintStep(prev, succ))
+  }
+}
+
+/**
+ * A data flow sink for Server Side Template Injection (SSTI) vulnerabilities
+ */
+abstract private class Sink extends DataFlow::ExprNode { }
+
+/**
+ * A data flow step for Server Side Template Injection (SSTI) vulnerabilities
+ */
+private class AdditionalFlowStep extends Unit {
+  abstract predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ);
+}
+
+/**
+ * An argument to FreeMarker template engine's `process` method call.
+ */
+private class FreeMarkerProcessSink extends Sink {
+  FreeMarkerProcessSink() {
+    exists(MethodAccess m |
+      m.getCallee() instanceof MethodFreeMarkerTemplateProcess and
+      m.getArgument(0) = this.getExpr()
+    )
+  }
+}
+
+/**
+ * An reader passed an argument to FreeMarker template engine's `Template`
+ *   construtor call.
+ */
+private class FreeMarkerConstructorSink extends Sink {
+  FreeMarkerConstructorSink() {
+    // Template(java.lang.String name, java.io.Reader reader)
+    // Template(java.lang.String name, java.io.Reader reader, Configuration cfg)
+    // Template(java.lang.String name, java.io.Reader reader, Configuration cfg, java.lang.String encoding)
+    // Template(java.lang.String name, java.lang.String sourceCode, Configuration cfg)
```

Comment:
```suggestion
```

Reply:
@smowton why should this be removed?

Reply:
Because it doesn't match this disjunct. This overload is handled in the disjunct below.